Mathlib/LinearAlgebra/TensorProduct/RightExactness.lean

```diff
@@ -0,0 +1,324 @@
+
+variable (g)
+
+lemma le_comap_range_lTensor (q : Q) :
+    LinearMap.range g ≤ (LinearMap.range (lTensor Q g)).comap (TensorProduct.mk R Q P q) := by
+  rintro x ⟨n, rfl⟩
+  exact ⟨q ⊗ₜ[R] n, rfl⟩
+
+lemma le_comap_range_rTensor (q : Q) :
+    LinearMap.range g ≤ (LinearMap.range (rTensor Q g)).comap
+      ((TensorProduct.mk R P Q).flip q) := by
+  rintro x ⟨n, rfl⟩
+  exact ⟨n ⊗ₜ[R] q, rfl⟩
+
+variable (Q) {g}
+
+theorem rTensor.surjective (hg : Function.Surjective g) :
+    Function.Surjective (rTensor Q g) := by
+  intro z
+  induction z using TensorProduct.induction_on with
+  | zero => exact ⟨0, map_zero _⟩
+  | tmul p q =>
+    obtain ⟨n, rfl⟩ := hg p
+    exact ⟨n ⊗ₜ[R] q, rfl⟩
+  | add x y hx hy =>
+    obtain ⟨x, rfl⟩ := hx
+    obtain ⟨y, rfl⟩ := hy
+    exact ⟨x + y, map_add _ _ _⟩
+
+theorem lTensor.surjective (hg : Function.Surjective g) :
+    Function.Surjective (lTensor Q g) := by
+  intro z
+  induction z using TensorProduct.induction_on with
+  | zero => exact ⟨0, map_zero _⟩
+  | tmul q p =>
+    obtain ⟨n, rfl⟩ := hg p
+    exact ⟨q ⊗ₜ[R] n, rfl⟩
+  | add x y hx hy =>
+    obtain ⟨x, rfl⟩ := hx
+    obtain ⟨y, rfl⟩ := hy
+    exact ⟨x + y, map_add _ _ _⟩
+
+end Semiring
+
+variable {R : Type*} [CommRing R]
+variable {M N P P' : Type*}
+  [AddCommGroup M] [AddCommGroup N] [AddCommGroup P]
+  [Module R M] [Module R N] [Module R P]
+
+lemma LinearMap.exact_subtype_mkQ (N : Submodule R N) :
+    Exact (Submodule.subtype N) (Submodule.mkQ N) := by
+  rw [LinearMap.exact_iff, Submodule.ker_mkQ, Submodule.range_subtype N]
+
+lemma LinearMap.exact_map_mkQ_range (f : M →ₗ[R] N) :
+    Exact f (Submodule.mkQ (range f)) :=
+  LinearMap.exact_iff.mpr <| Submodule.ker_mkQ _
+
+lemma LinearMap.exact_subtype_ker_map (g : N →ₗ[R] P) :
+    Exact (Submodule.subtype (ker g)) g :=
+  LinearMap.exact_iff.mpr <| (Submodule.range_subtype _).symm
+
+variable {f : M →ₗ[R] N} {g : N →ₗ[R] P}
+
+variable (Q : Type*) [AddCommGroup Q] [Module R Q]
+
+variable (hfg : Exact f g) (hg : Function.Surjective g)
+
+-- private
+def rTensor.inverse_ofRightInverse {h : P → N} (hgh : Function.RightInverse h g) :
+    P ⊗[R] Q →ₗ[R] N ⊗[R] Q ⧸ LinearMap.range (rTensor Q f) := by
+  rw [exact_iff] at hfg
+  apply TensorProduct.lift
+  apply LinearMap.mk₂ R (fun p q ↦ Submodule.Quotient.mk (h p ⊗ₜ[R] q))
+  · intro p p' q
+    rw [← Submodule.Quotient.mk_add, Submodule.Quotient.eq]
+    rw [← TensorProduct.add_tmul, ← TensorProduct.sub_tmul]
+    apply le_comap_range_rTensor f
+    rw [← hfg, mem_ker, map_sub, map_add]
+    simp only [hgh _, sub_self]
+  · intro r p q
+    rw [← Submodule.Quotient.mk_smul, Submodule.Quotient.eq]
+    rw [TensorProduct.smul_tmul', ← TensorProduct.sub_tmul]
+    apply le_comap_range_rTensor f
+    rw [← hfg, mem_ker, map_sub, map_smul]
+    simp only [hgh _, sub_self]
+  · intro p q q'
+    rw [TensorProduct.tmul_add, ← Submodule.Quotient.mk_add]
+  · intro r p q
+    simp only [TensorProduct.tmul_smul, Submodule.Quotient.mk_smul]
```

Comment:
What do you think of:
```suggestion
  rw [exact_iff] at hfg
  refine
    TensorProduct.lift
      { toFun := fun p ↦ Submodule.mkQ _ ∘ₗ TensorProduct.mk R _ _ (h p)
        map_add' := fun p p' => LinearMap.ext <| fun q => (Submodule.Quotient.eq _).mpr ?_
        map_smul' := fun r p => LinearMap.ext <| fun q => (Submodule.Quotient.eq _).mpr ?_ }
  · change h (p + p') ⊗ₜ[R] q - (h p ⊗ₜ[R] q + h p' ⊗ₜ[R] q) ∈ range (rTensor Q f)
    rw [← TensorProduct.add_tmul, ← TensorProduct.sub_tmul]
    apply le_comap_range_rTensor f
    rw [← hfg, mem_ker, map_sub, map_add]
    simp only [hgh _, sub_self]
  · change h (r • p) ⊗ₜ[R] q - r • h p ⊗ₜ[R] q ∈ range (rTensor Q f)
    rw [TensorProduct.smul_tmul', ← TensorProduct.sub_tmul]
    apply le_comap_range_rTensor f
    rw [← hfg, mem_ker, map_sub, map_smul]
    simp only [hgh _, sub_self]
```
I think this makes it a bit clearer that half of the linearity is obvious, while the other half needs exactness

Reply:
For me, it's harder to read, but OK.

Reply:
Do you want me to update the second definition in the same way?

Reply:
I'm happy to leave this for another reviewer to tie break if you don't like it.

Reply:
That's OK. That's actually good to be able to adjust to other people's stylistic habits.

Reply:
I made the effort of writing the other proof — that helped me appreciate this style.